File: apps/mobile/lib/services/websocket_service.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import '../models/session.dart';

/// Connection state for WebSocket
enum WsConnectionState {
  disconnected,
  connecting,
  connected,
  error,
}

/// Service for managing WebSocket connection to TUI Parser
class WebSocketService extends ChangeNotifier {
  WebSocketChannel? _channel;
  StreamSubscription? _subscription;

  WsConnectionState _connectionState = WsConnectionState.disconnected;
  String? _errorMessage;
  String _serverAddress = 'localhost:9600';

  // Session data
  final List<Session> _sessions = [];
  String? _currentSessionId;
  final Map<String, List<OutputLine>> _sessionOutputs = {};
  Prompt? _activePrompt;
  Menu? _activeMenu;

  // Getters
  WsConnectionState get connectionState => _connectionState;
  String? get errorMessage => _errorMessage;
  String get serverAddress => _serverAddress;
  List<Session> get sessions => List.unmodifiable(_sessions);
  String? get currentSessionId => _currentSessionId;
  List<OutputLine> get currentOutput =>
      _sessionOutputs[_currentSessionId] ?? [];
  Prompt? get activePrompt => _activePrompt;
  Menu? get activeMenu => _activeMenu;

  /// Connect to the TUI Parser server
  Future<void> connect(String address) async {
    _serverAddress = address;
    _connectionState = WsConnectionState.connecting;
    _errorMessage = null;
    notifyListeners();

    try {
      final uri = Uri.parse('ws://$address/ws');
      _channel = WebSocketChannel.connect(uri);

      _subscription = _channel!.stream.listen(
        _handleMessage,
        onError: _handleError,
        onDone: _handleDone,
      );

      _connectionState = WsConnectionState.connected;
      notifyListeners();
    } catch (e) {
      _connectionState = WsConnectionState.error;
      _errorMessage = e.toString();
      notifyListeners();
    }
  }

  /// Disconnect from the server
  void disconnect() {
    _subscription?.cancel();
    _channel?.sink.close();
    _channel = null;
    _connectionState = WsConnectionState.disconnected;
    notifyListeners();
  }

  /// Subscribe to a session
  void subscribe(String sessionId) {
    _sendMessage({
      'type': 'subscribe',
      'sessionIds': [sessionId],
    });
    _currentSessionId = sessionId;
    _sessionOutputs.putIfAbsent(sessionId, () => []);
    notifyListeners();
  }

  /// Unsubscribe from a session
  void unsubscribe(String sessionId) {
    _sendMessage({
      'type': 'unsubscribe',
      'sessionIds': [sessionId],
    });
    if (_currentSessionId == sessionId) {
      _currentSessionId = null;
    }
    notifyListeners();
  }

  /// Send text input to current session
  void sendInput(String text) {
    if (_currentSessionId == null) return;
    _sendMessage({
      'type': 'input',
      'sessionId': _currentSessionId,
      'text': text,
    });
  }

  /// Send a special key to current session
  void sendKey(String key) {
    if (_currentSessionId == null) return;
    _sendMessage({
      'type': 'key',
      'sessionId': _currentSessionId,
      'key': key,
    });
  }

  /// Send an action response (for prompts/menus)
  void sendAction(String action, [String? value]) {
    if (_currentSessionId == null) return;
    _sendMessage({
      'type': 'action',
      'sessionId': _currentSessionId,
      'action': action,
      'value': value,
    });
    // Clear active prompt/menu after action
    _activePrompt = null;
    _activeMenu = null;
    notifyListeners();
  }

  /// Clear output for a session
  void clearOutput(String sessionId) {
    _sessionOutputs[sessionId]?.clear();
    notifyListeners();
  }

  void _sendMessage(Map<String, dynamic> message) {
    if (_channel == null) return;
    _channel!.sink.add(jsonEncode(message));
  }

  void _handleMessage(dynamic message) {
    try {
      final data = jsonDecode(message as String) as Map<String, dynamic>;
      final type = data['type'] as String?;

      switch (type) {
        case 'sessions':
          _handleSessions(data);
          break;
        case 'output':
          _handleOutput(data);
          break;
        case 'prompt':
          _handlePrompt(data);
          break;
        case 'menu':
          _handleMenu(data);
          break;
        case 'status':
          _handleStatus(data);
          break;
        case 'clear':
          _handleClear(data);
          break;
        case 'error':
          _handleServerError(data);
          break;
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error parsing message: $e');
    }
  }

  void _handleSessions(Map<String, dynamic> data) {
    final sessionList = data['sessions'] as List<dynamic>?;
    _sessions.clear();
    if (sessionList != null) {
      for (final s in sessionList) {
        _sessions.add(Session.fromJson(s as Map<String, dynamic>));
      }
    }
  }

  void _handleOutput(Map<String, dynamic> data) {
    final sessionId = data['sessionId'] as String?;
    final text = data['text'] as String?;
    if (sessionId == null || text == null) return;

    _sessionOutputs.putIfAbsent(sessionId, () => []);

    // Split text into lines and add each
    final lines = text.split('\n');
    for (int i = 0; i < lines.length; i++) {
      final line = lines[i];
      // Only add non-empty lines, or add empty line if not the last split
      if (line.isNotEmpty || i < lines.length - 1) {
        _sessionOutputs[sessionId]!.add(OutputLine(text: line));
      }
    }

    // Limit output buffer to 10000 lines
    while (_sessionOutputs[sessionId]!.length > 10000) {
      _sessionOutputs[sessionId]!.removeAt(0);
    }
  }

  void _handlePrompt(Map<String, dynamic> data) {
    _activePrompt = Prompt.fromJson(data);
    _activeMenu = null;
  }

  void _handleMenu(Map<String, dynamic> data) {
    _activeMenu = Menu.fromJson(data);
    _activePrompt = null;
  }

  void _handleStatus(Map<String, dynamic> data) {
    final sessionId = data['sessionId'] as String?;
    final status = data['status'] as String?;
    if (sessionId == null || status == null) return;

    // Update session status
    final index = _sessions.indexWhere((s) => s.id == sessionId);
    if (index >= 0) {
      _sessions[index] = Session(
        id: _sessions[index].id,
        name: _sessions[index].name,
        status: status,
        lastActivity: DateTime.now(),
      );
    }
  }

  void _handleClear(Map<String, dynamic> data) {
    final sessionId = data['sessionId'] as String?;
    if (sessionId != null) {
      _sessionOutputs[sessionId]?.clear();
    }
  }

  void _handleServerError(Map<String, dynamic> data) {
    final error = data['error'] as String?;
    if (error != null) {
      _errorMessage = error;
    }
  }

  void _handleError(dynamic error) {
    _connectionState = WsConnectionState.error;
    _errorMessage = error.toString();
    notifyListeners();
  }

  void _handleDone() {
    _connectionState = WsConnectionState.disconnected;
    notifyListeners();
  }

  @override
  void dispose() {
    disconnect();
    super.dispose();
  }
}

File: apps/ios/Sources/PlexusOneiOS/Services/WebSocketService.swift
import Foundation
import Combine

/// Connection state for WebSocket
enum ConnectionState {
    case disconnected
    case connecting
    case connected
    case error(String)
}

/// WebSocket service for communicating with tuiparser
@MainActor
class WebSocketService: ObservableObject {
    @Published var connectionState: ConnectionState = .disconnected
    @Published var sessions: [Session] = []
    @Published var currentSessionId: String?
    @Published var currentOutput: String = ""
    @Published var debugLog: String = ""  // Debug: shows last messages

    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private var serverAddress: String = "ws://127.0.0.1:9600/ws"

    /// Output buffer per session
    private var outputBuffers: [String: String] = [:]

    init() {}

    /// Connect to the WebSocket server
    func connect(to address: String? = nil) {
        if let address = address {
            serverAddress = address
        }

        print("[WebSocket] Connecting to: \(serverAddress)")
        disconnect()
        connectionState = .connecting

        guard let url = URL(string: serverAddress) else {
            connectionState = .error("Invalid URL: \(serverAddress)")
            return
        }

        urlSession = URLSession(configuration: .default)
        webSocketTask = urlSession?.webSocketTask(with: url)
        webSocketTask?.resume()

        connectionState = .connected
        debugLog = "Connected to \(serverAddress)\n" + debugLog
        print("[WebSocket] Connected, starting receive loop")
        receiveMessage()

        // Request session list
        debugLog = "Sending list request\n" + debugLog
        print("[WebSocket] Sending list request")
        sendMessage(["type": "list"])
    }

    /// Disconnect from the server
    func disconnect() {
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        urlSession = nil
        connectionState = .disconnected
    }

    /// Subscribe to a session's output
    func subscribe(to sessionId: String) {
        currentSessionId = sessionId
        currentOutput = outputBuffers[sessionId] ?? ""
        sendMessage([
            "type": "subscribe",
            "sessionIds": [sessionId]  // Server expects array
        ])
    }

    /// Send an action (for prompts)
    func sendAction(_ action: String) {
        guard let sessionId = currentSessionId else { return }
        sendMessage([
            "type": "action",
            "sessionId": sessionId,
            "action": action
        ])
    }

    /// Send text input
    func sendInput(_ text: String) {
        guard let sessionId = currentSessionId else { return }
        sendMessage([
            "type": "input",
            "sessionId": sessionId,
            "text": text
        ])
    }

    /// Send a key press
    func sendKey(_ key: String) {
        guard let sessionId = currentSessionId else { return }
        sendMessage([
            "type": "key",
            "sessionId": sessionId,
            "key": key
        ])
    }

    /// Update server address
    func setServerAddress(_ address: String) {
        serverAddress = address
    }

    /// Get current server address
    var currentServerAddress: String {
        serverAddress
    }

    // MARK: - Private Methods

    private func sendMessage(_ dict: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: dict),
              let string = String(data: data, encoding: .utf8) else {
            return
        }

        webSocketTask?.send(.string(string)) { [weak self] error in
            if let error = error {
                Task { @MainActor in
                    self?.connectionState = .error(error.localizedDescription)
                }
            }
        }
    }

    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            Task { @MainActor in
                switch result {
                case .success(let message):
                    switch message {
                    case .string(let text):
                        self?.handleMessage(text)
                    case .data(let data):
                        if let text = String(data: data, encoding: .utf8) {
                            self?.handleMessage(text)
                        }
                    @unknown default:
                        break
                    }
                    // Continue receiving
                    self?.receiveMessage()

                case .failure(let error):
                    self?.debugLog = "RECV ERROR: \(error.localizedDescription)\n" + (self?.debugLog ?? "")
                    self?.connectionState = .error(error.localizedDescription)
                }
            }
        }
    }

    private func handleMessage(_ text: String) {
        let preview = String(text.prefix(100))
        debugLog = "Recv: \(preview)...\n" + debugLog.prefix(500)
        print("[WebSocket] Received message: \(text.prefix(200))...")

        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let typeString = json["type"] as? String else {
            debugLog = "PARSE FAIL\n" + debugLog
            print("[WebSocket] Failed to parse message as JSON")
            return
        }

        debugLog = "Type: \(typeString)\n" + debugLog
        print("[WebSocket] Message type: \(typeString)")

        switch typeString {
        case "sessions":
            handleSessionsMessage(json)
        case "output":
            handleOutputMessage(json)
        case "status":
            handleStatusMessage(json)
        case "error":
            handleErrorMessage(json)
        default:
            print("[WebSocket] Unknown message type: \(typeString)")
            break
        }
    }

    private func handleSessionsMessage(_ json: [String: Any]) {
        debugLog = "handleSessions called\n" + debugLog
        print("[WebSocket] Handling sessions message")
        guard let sessionsData = json["sessions"] as? [[String: Any]] else {
            debugLog = "SESSIONS PARSE FAIL - keys: \(Array(json.keys))\n" + debugLog
            print("[WebSocket] Failed to parse 'sessions' array from message")
            print("[WebSocket] JSON keys: \(json.keys)")
            return
        }

        debugLog = "Found \(sessionsData.count) raw sessions\n" + debugLog
        print("[WebSocket] Found \(sessionsData.count) sessions in response")

        sessions = sessionsData.compactMap { dict in
            guard let id = dict["id"] as? String,
                  let name = dict["name"] as? String else {
                debugLog = "Session parse fail: \(dict.keys)\n" + debugLog
                print("[WebSocket] Failed to parse session: \(dict)")
                return nil
            }
            let statusString = dict["status"] as? String ?? "unknown"
            let status = Session.SessionStatus(rawValue: statusString) ?? .unknown
            return Session(id: id, name: name, status: status)
        }

        debugLog = "Parsed \(sessions.count) sessions\n" + debugLog
        print("[WebSocket] Parsed \(sessions.count) sessions")

        // Auto-select first session if none selected
        if currentSessionId == nil, let first = sessions.first {
            subscribe(to: first.id)
        }
    }

    private func handleOutputMessage(_ json: [String: Any]) {
        guard let sessionId = json["sessionId"] as? String,
              let output = json["text"] as? String else {  // Server sends "text", not "output"
            return
        }

        // Append to buffer
        var buffer = outputBuffers[sessionId] ?? ""
        buffer += output
        outputBuffers[sessionId] = buffer

        // Update current output if this is the active session
        if sessionId == currentSessionId {
            currentOutput = buffer
        }
    }

    private func handleStatusMessage(_ json: [String: Any]) {
        guard let sessionId = json["sessionId"] as? String,
              let statusString = json["status"] as? String,
              let status = Session.SessionStatus(rawValue: statusString) else {
            return
        }

        if let index = sessions.firstIndex(where: { $0.id == sessionId }) {
            let session = sessions[index]
            sessions[index] = Session(id: session.id, name: session.name, status: status)
        }
    }

    private func handleErrorMessage(_ json: [String: Any]) {
        if let message = json["message"] as? String {
            connectionState = .error(message)
        }
    }

    /// Clear output buffer for a session
    func clearOutput(for sessionId: String? = nil) {
        let id = sessionId ?? currentSessionId
        guard let id = id else { return }
        outputBuffers[id] = ""
        if id == currentSessionId {
            currentOutput = ""
        }
    }

    /// Refresh session list
    func refreshSessions() {
        sendMessage(["type": "list"])
    }

    /// Create a new tmux session
    func createSession(name: String) {
        sendMessage([
            "type": "create",
            "name": name
        ])
        // Refresh session list after creation
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            self.refreshSessions()
        }
    }
}
